pkg/erofs/erofs.go

```diff
@@ -690,87 +725,162 @@ func (i *Inode) blocks() uint64 {
 //
 // [ (metadata block) inode | optional fields | dirent M+2 | dirent M+3 | name M+2 | name M+3 | optional padding ]
 //
-// All directory entries are _strictly_ recorded in alphabetical order.
-func (i *Inode) IterDirents(cb func(name string, typ uint8, nid uint64) error) error {
+// refer: https://docs.kernel.org/filesystems/erofs.html#directories
+func (i *Inode) getDirentName(d *Dirent, block blockData, lastDirent bool) ([]byte, *Dirent, error) {
+	var (
+		next    *Dirent
+		nameLen uint32
+	)
+	if lastDirent {
+		nameLen = block.size - uint32(d.NameOff)
+	} else {
+		var err error
+		if next, err = i.image.direntAfter(d); err != nil {
+			return nil, nil, err
+		}
+		nameLen = uint32(next.NameOff - d.NameOff)
+	}
+	if uint32(d.NameOff)+nameLen > block.size || nameLen > MaxNameLen || nameLen == 0 {
+		log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+		return nil, nil, linuxerr.EUCLEAN
+	}
+	name, err := i.image.BytesAt(block.base+uint64(d.NameOff), uint64(nameLen))
+	if err != nil {
+		return nil, nil, err
+	}
+	if lastDirent {
+		// Optional padding may exist at the end of a block.
+		n := bytes.IndexByte(name, 0)
+		if n == 0 {
+			log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+			return nil, nil, linuxerr.EUCLEAN
+		}
+		if n != -1 {
+			name = name[:n]
+		}
+	}
+	return name, next, nil
+}
+
+// getDirent0 returns a pointer to the first dirent and the number of dirents
+// in the given block in this inode.
+func (i *Inode) getDirent0(block blockData) (*Dirent, uint16, error) {
+	d0, err := i.image.direntAt(block.base)
+	if err != nil {
+		return nil, 0, err
+	}
+	// Apart from the offset of the first filename, nameOff0 also indicates
+	// the total number of dirents in this block.
+	if d0.NameOff < DirentSize || uint32(d0.NameOff) >= block.size {
+		log.Warningf("Invalid nameOff0 %v at inode (nid=%v)", d0.NameOff, i.Nid())
+		return nil, 0, linuxerr.EUCLEAN
+	}
+	numDirents := d0.NameOff / DirentSize
+	return d0, numDirents, nil
+}
+
+// Lookup looks up a child by the name. The child inode number will be returned on success.
+func (i *Inode) Lookup(name string) (uint64, error) {
 	if !i.IsDir() {
-		return linuxerr.ENOTDIR
+		return 0, linuxerr.ENOTDIR
 	}
 
-	blocks := i.blocks()
-	blockSize := i.image.BlockSize()
+	nameBytes := gohacks.ImmutableBytesFromString(name)
```

Comment:
It is mainly to allow us to efficiently do three-way comparisons between names, because using three-way comparison is beneficial for the readability and performance (e.g. fewer iterations) in this case, and it seems there is no efficient way to do three-way comparisons between strings in Go currently.

I tried below code with go1.21.3, and the compiler will generate two function calls (`runtime.memequal` and `runtime.cmpstring`, i.e. the strings may be compared twice):

```go
package main

func cmp(a string, b []byte) int {
	switch {
	case a == string(b):
		return 0
	case a > string(b):
		return 1
	default:
		return -1
	}
}
```

```
main.cmp STEXT size=197 args=0x28 locals=0x28 funcid=0x0 align=0x0
	0x0000 00000 (/home/test/tmp.go:3)	TEXT	main.cmp(SB), ABIInternal, $40-40
	0x0000 00000 (/home/test/tmp.go:3)	CMPQ	SP, 16(R14)
	0x0004 00004 (/home/test/tmp.go:3)	PCDATA	$0, $-2
	0x0004 00004 (/home/test/tmp.go:3)	JLS	133
	0x0006 00006 (/home/test/tmp.go:3)	PCDATA	$0, $-1
	0x0006 00006 (/home/test/tmp.go:3)	PUSHQ	BP
	0x0007 00007 (/home/test/tmp.go:3)	MOVQ	SP, BP
	0x000a 00010 (/home/test/tmp.go:3)	SUBQ	$32, SP
	0x000e 00014 (/home/test/tmp.go:3)	MOVQ	AX, main.a+48(FP)
	0x0013 00019 (/home/test/tmp.go:3)	MOVQ	CX, main.b+64(FP)
	0x0018 00024 (/home/test/tmp.go:3)	FUNCDATA	$0, gclocals·iilYh2zWk/RieCMyRG2Y4w==(SB)
	0x0018 00024 (/home/test/tmp.go:3)	FUNCDATA	$1, gclocals·J5F+7Qw7O7ve2QcWC7DpeQ==(SB)
	0x0018 00024 (/home/test/tmp.go:3)	FUNCDATA	$5, main.cmp.arginfo1(SB)
	0x0018 00024 (/home/test/tmp.go:3)	FUNCDATA	$6, main.cmp.argliveinfo(SB)
	0x0018 00024 (/home/test/tmp.go:3)	PCDATA	$3, $1
	0x0018 00024 (/home/test/tmp.go:5)	CMPQ	DI, BX
	0x001b 00027 (/home/test/tmp.go:5)	JNE	88
	0x001d 00029 (/home/test/tmp.go:5)	MOVQ	CX, main.b+64(SP)
	0x0022 00034 (/home/test/tmp.go:5)	MOVQ	DI, main.b+72(SP)
	0x0027 00039 (/home/test/tmp.go:3)	MOVQ	BX, main.a+56(SP)
	0x002c 00044 (/home/test/tmp.go:3)	MOVQ	AX, main.a+48(SP)
	0x0031 00049 (/home/test/tmp.go:3)	PCDATA	$3, $2
	0x0031 00049 (/home/test/tmp.go:3)	MOVQ	BX, DX
	0x0034 00052 (/home/test/tmp.go:5)	MOVQ	CX, BX
	0x0037 00055 (/home/test/tmp.go:5)	MOVQ	DX, CX
	0x003a 00058 (/home/test/tmp.go:5)	PCDATA	$1, $0
	0x003a 00058 (/home/test/tmp.go:5)	CALL	runtime.memequal(SB)
	0x003f 00063 (/home/test/tmp.go:5)	NOP
	0x0040 00064 (/home/test/tmp.go:5)	TESTB	AL, AL
	0x0042 00066 (/home/test/tmp.go:5)	JNE	125
	0x0044 00068 (/home/test/tmp.go:7)	MOVQ	main.a+48(SP), AX
	0x0049 00073 (/home/test/tmp.go:7)	MOVQ	main.b+64(SP), CX
	0x004e 00078 (/home/test/tmp.go:7)	MOVQ	main.a+56(SP), BX
	0x0053 00083 (/home/test/tmp.go:7)	MOVQ	main.b+72(SP), DI
	0x0058 00088 (/home/test/tmp.go:5)	PCDATA	$1, $-1
	0x0058 00088 (/home/test/tmp.go:5)	PCDATA	$3, $1
	0x0058 00088 (/home/test/tmp.go:7)	PCDATA	$1, $1
	0x0058 00088 (/home/test/tmp.go:7)	CALL	runtime.cmpstring(SB)
	0x005d 00093 (/home/test/tmp.go:7)	NOP
	0x0060 00096 (/home/test/tmp.go:7)	TESTQ	AX, AX
	0x0063 00099 (/home/test/tmp.go:7)	JLE	112
	0x0065 00101 (/home/test/tmp.go:8)	MOVL	$1, AX
	0x006a 00106 (/home/test/tmp.go:8)	ADDQ	$32, SP
	0x006e 00110 (/home/test/tmp.go:8)	POPQ	BP
	0x006f 00111 (/home/test/tmp.go:8)	RET
	0x0070 00112 (/home/test/tmp.go:10)	MOVQ	$-1, AX
	0x0077 00119 (/home/test/tmp.go:10)	ADDQ	$32, SP
	0x007b 00123 (/home/test/tmp.go:10)	POPQ	BP
	0x007c 00124 (/home/test/tmp.go:10)	RET
	0x007d 00125 (/home/test/tmp.go:10)	PCDATA	$1, $-1
	0x007d 00125 (/home/test/tmp.go:10)	PCDATA	$3, $2
	0x007d 00125 (/home/test/tmp.go:6)	XORL	AX, AX
	0x007f 00127 (/home/test/tmp.go:6)	ADDQ	$32, SP
	0x0083 00131 (/home/test/tmp.go:6)	POPQ	BP
	0x0084 00132 (/home/test/tmp.go:6)	RET
	0x0085 00133 (/home/test/tmp.go:6)	NOP
	0x0085 00133 (/home/test/tmp.go:3)	PCDATA	$1, $-1
	0x0085 00133 (/home/test/tmp.go:3)	PCDATA	$0, $-2
	0x0085 00133 (/home/test/tmp.go:3)	MOVQ	AX, 8(SP)
	0x008a 00138 (/home/test/tmp.go:3)	MOVQ	BX, 16(SP)
	0x008f 00143 (/home/test/tmp.go:3)	MOVQ	CX, 24(SP)
	0x0094 00148 (/home/test/tmp.go:3)	MOVQ	DI, 32(SP)
	0x0099 00153 (/home/test/tmp.go:3)	MOVQ	SI, 40(SP)
	0x009e 00158 (/home/test/tmp.go:3)	NOP
	0x00a0 00160 (/home/test/tmp.go:3)	CALL	runtime.morestack_noctxt(SB)
	0x00a5 00165 (/home/test/tmp.go:3)	MOVQ	8(SP), AX
	0x00aa 00170 (/home/test/tmp.go:3)	MOVQ	16(SP), BX
	0x00af 00175 (/home/test/tmp.go:3)	MOVQ	24(SP), CX
	0x00b4 00180 (/home/test/tmp.go:3)	MOVQ	32(SP), DI
	0x00b9 00185 (/home/test/tmp.go:3)	MOVQ	40(SP), SI
	0x00be 00190 (/home/test/tmp.go:3)	PCDATA	$0, $-1
	0x00be 00190 (/home/test/tmp.go:3)	NOP
	0x00c0 00192 (/home/test/tmp.go:3)	JMP	0
```

Reply:
> it seems there is no efficient way to do three-way comparisons between strings in Go currently

`strings.Compare()`?

Reply:
Yeah, I also considered `strings.Compare()`. There are two issues when using `strings.Compare()`:

- there will be an extra copy from `[]byte` to `string` (`runtime.slicebytetostring`);
- strings may be compared twice (`runtime.memequal` and `runtime.cmpstring`);

Below is the test code:

```go
package main

import "strings"

func cmp(a string, b []byte) int {
	return strings.Compare(a, string(b))
}
```

And this is the code generated by the compiler:

```
main.cmp STEXT size=213 args=0x28 locals=0x58 funcid=0x0 align=0x0
	0x0000 00000 (/home/test/tmp.go:5)	TEXT	main.cmp(SB), ABIInternal, $88-40
	0x0000 00000 (/home/test/tmp.go:5)	CMPQ	SP, 16(R14)
	0x0004 00004 (/home/test/tmp.go:5)	PCDATA	$0, $-2
	0x0004 00004 (/home/test/tmp.go:5)	JLS	153
	0x000a 00010 (/home/test/tmp.go:5)	PCDATA	$0, $-1
	0x000a 00010 (/home/test/tmp.go:5)	PUSHQ	BP
	0x000b 00011 (/home/test/tmp.go:5)	MOVQ	SP, BP
	0x000e 00014 (/home/test/tmp.go:5)	SUBQ	$80, SP
	0x0012 00018 (/home/test/tmp.go:5)	MOVQ	CX, main.b+112(FP)
	0x0017 00023 (/home/test/tmp.go:5)	FUNCDATA	$0, gclocals·vK+vs0a3g9FAwMMfjFVdug==(SB)
	0x0017 00023 (/home/test/tmp.go:5)	FUNCDATA	$1, gclocals·ESr+dlj+ZUXO8X1Jig2pHQ==(SB)
	0x0017 00023 (/home/test/tmp.go:5)	FUNCDATA	$5, main.cmp.arginfo1(SB)
	0x0017 00023 (/home/test/tmp.go:5)	FUNCDATA	$6, main.cmp.argliveinfo(SB)
	0x0017 00023 (/home/test/tmp.go:5)	PCDATA	$3, $1
	0x0017 00023 (/home/test/tmp.go:5)	MOVQ	BX, main.a+104(SP)
	0x001c 00028 (/home/test/tmp.go:5)	MOVQ	AX, main.a+96(SP)
	0x0021 00033 (/home/test/tmp.go:5)	PCDATA	$3, $2
	0x0021 00033 (/home/test/tmp.go:6)	LEAQ	main..autotmp_6+40(SP), AX
	0x0026 00038 (/home/test/tmp.go:6)	MOVQ	CX, BX
	0x0029 00041 (/home/test/tmp.go:6)	MOVQ	DI, CX
	0x002c 00044 (/home/test/tmp.go:6)	PCDATA	$1, $1
	0x002c 00044 (/home/test/tmp.go:6)	CALL	runtime.slicebytetostring(SB)
	0x0031 00049 (<unknown line number>)	NOP
	0x0031 00049 (/home/test/go1.21.3/src/strings/compare.go:21)	MOVQ	main.a+104(SP), CX
	0x0036 00054 (/home/test/go1.21.3/src/strings/compare.go:21)	CMPQ	BX, CX
	0x0039 00057 (/home/test/go1.21.3/src/strings/compare.go:21)	JNE	105
	0x003b 00059 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	AX, main.b.ptr+72(SP)
	0x0040 00064 (/home/test/go1.21.3/src/strings/compare.go:21)	MOVQ	BX, main.b.len+32(SP)
	0x0045 00069 (/home/test/go1.21.3/src/strings/compare.go:21)	MOVQ	AX, BX
	0x0048 00072 (/home/test/go1.21.3/src/strings/compare.go:21)	MOVQ	main.a+96(SP), AX
	0x004d 00077 (/home/test/go1.21.3/src/strings/compare.go:21)	PCDATA	$1, $2
	0x004d 00077 (/home/test/go1.21.3/src/strings/compare.go:21)	CALL	runtime.memequal(SB)
	0x0052 00082 (/home/test/go1.21.3/src/strings/compare.go:21)	TESTB	AL, AL
	0x0054 00084 (/home/test/go1.21.3/src/strings/compare.go:21)	JEQ	90
	0x0056 00086 (/home/test/go1.21.3/src/strings/compare.go:21)	XORL	AX, AX
	0x0058 00088 (/home/test/tmp.go:6)	JMP	147
	0x005a 00090 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	main.b.ptr+72(SP), AX
	0x005f 00095 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	main.a+104(SP), CX
	0x0064 00100 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	main.b.len+32(SP), BX
	0x0069 00105 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	BX, DI
	0x006c 00108 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	CX, BX
	0x006f 00111 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	AX, CX
	0x0072 00114 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	main.a+96(SP), AX
	0x0077 00119 (/home/test/go1.21.3/src/strings/compare.go:24)	PCDATA	$1, $3
	0x0077 00119 (/home/test/go1.21.3/src/strings/compare.go:24)	CALL	runtime.cmpstring(SB)
	0x007c 00124 (/home/test/go1.21.3/src/strings/compare.go:24)	NOP
	0x0080 00128 (/home/test/go1.21.3/src/strings/compare.go:24)	TESTQ	AX, AX
	0x0083 00131 (/home/test/go1.21.3/src/strings/compare.go:24)	JGE	142
	0x0085 00133 (/home/test/go1.21.3/src/strings/compare.go:24)	MOVQ	$-1, AX
	0x008c 00140 (/home/test/tmp.go:6)	JMP	147
	0x008e 00142 (/home/test/tmp.go:6)	MOVL	$1, AX
	0x0093 00147 (/home/test/tmp.go:6)	PCDATA	$1, $-1
	0x0093 00147 (/home/test/tmp.go:6)	ADDQ	$80, SP
	0x0097 00151 (/home/test/tmp.go:6)	POPQ	BP
	0x0098 00152 (/home/test/tmp.go:6)	RET
	0x0099 00153 (/home/test/tmp.go:6)	NOP
	0x0099 00153 (/home/test/tmp.go:5)	PCDATA	$1, $-1
	0x0099 00153 (/home/test/tmp.go:5)	PCDATA	$0, $-2
	0x0099 00153 (/home/test/tmp.go:5)	MOVQ	AX, 8(SP)
	0x009e 00158 (/home/test/tmp.go:5)	MOVQ	BX, 16(SP)
	0x00a3 00163 (/home/test/tmp.go:5)	MOVQ	CX, 24(SP)
	0x00a8 00168 (/home/test/tmp.go:5)	MOVQ	DI, 32(SP)
	0x00ad 00173 (/home/test/tmp.go:5)	MOVQ	SI, 40(SP)
	0x00b2 00178 (/home/test/tmp.go:5)	CALL	runtime.morestack_noctxt(SB)
	0x00b7 00183 (/home/test/tmp.go:5)	MOVQ	8(SP), AX
	0x00bc 00188 (/home/test/tmp.go:5)	MOVQ	16(SP), BX
	0x00c1 00193 (/home/test/tmp.go:5)	MOVQ	24(SP), CX
	0x00c6 00198 (/home/test/tmp.go:5)	MOVQ	32(SP), DI
	0x00cb 00203 (/home/test/tmp.go:5)	MOVQ	40(SP), SI
	0x00d0 00208 (/home/test/tmp.go:5)	PCDATA	$0, $-1
	0x00d0 00208 (/home/test/tmp.go:5)	JMP	0
```

And based on the comments in `strings.Compare()`, the author doesn't want us to use this function:

https://github.com/golang/go/blob/dc5a0d276bbcd6120325c9e0f9c8fd099fa2b8d6/src/strings/compare.go#L10-L20

Reply:
Ah, thanks for the pointers. Lets leave this as is, maybe add a comment why we take all the effort to use gohacks (because we want to do efficient string three-way comparison).

Reply:
Sure, will do.